packages/arb-bridge-peripherals/contracts/tokenbridge/ethereum/gateway/L1GatewayRouter.sol

```diff
@@ -218,29 +218,58 @@ contract L1GatewayRouter is WhitelistConsumer, L1ArbitrumMessenger, GatewayRoute
             _setGateways(_token, _gateway, _maxGas, _gasPriceBid, _maxSubmissionCost, msg.sender);
     }
 
-    function outboundTransfer(
-        address _token,
-        address _to,
-        uint256 _amount,
+    function _outboundTransferChecks(
         uint256 _maxGas,
         uint256 _gasPriceBid,
-        bytes calldata _data
-    ) public payable override onlyWhitelisted returns (bytes memory) {
+        bytes memory _data
+    ) internal {
         // when sending a L1 to L2 transaction, we expect the user to send
         // eth in flight in order to pay for L2 gas costs
         // this check prevents users from misconfiguring the msg.value
-        (uint256 _maxSubmissionCost, ) = abi.decode(_data, (uint256, bytes));
+        uint256 _maxSubmissionCost;
+        // assembly code block below is the gas optimized version of
+        // _maxSubmissionCost = abi.decode(_data, (uint256, bytes));
+        assembly {
+            _maxSubmissionCost := mload(add(_data, 0x20))
+        }
 
         // here we don't use SafeMath since this validation is to prevent users
         // from shooting themselves on the foot.
-        uint256 expectedEth = _maxSubmissionCost + (_maxGas * _gasPriceBid);
-        require(_maxSubmissionCost > 0, "NO_SUBMISSION_COST");
-        require(msg.value == expectedEth, "WRONG_ETH_VALUE");
+        require(_maxSubmissionCost != 0, "NO_SUBMISSION_COST");
+        require(msg.value == _maxSubmissionCost + (_maxGas * _gasPriceBid), "WRONG_ETH_VALUE");
+    }
+
+    function outboundTransfer(
+        address _token,
+        address _to,
+        uint256 _amount,
+        uint256 _maxGas,
+        uint256 _gasPriceBid,
+        bytes calldata _data
+    ) public payable override onlyWhitelisted returns (bytes memory) {
+        _outboundTransferChecks(_maxGas, _gasPriceBid, _data);
 
         // will revert if msg.sender is not whitelisted
         return super.outboundTransfer(_token, _to, _amount, _maxGas, _gasPriceBid, _data);
     }
 
+    function outboundTransferCustomRefund(
+        address _token,
+        address _refundTo,
+        address _to,
+        uint256 _amount,
+        uint256 _maxGas,
+        uint256 _gasPriceBid,
+        bytes calldata _data
+    ) public payable override onlyWhitelisted returns (bytes memory) {
+        // _refundTo is subject to L2 alias rewrite
```

Comment:
this is important, we should be add it as a natspec comment to the function

Reply:
added some natspec comments